builder/role_image_test.go

```diff
@@ -210,94 +211,125 @@ func TestGenerateRoleImageDockerfileDir(t *testing.T) {
+		"root/var/vcap/jobs-src/tor/config_spec.json":             {desc: "tor config spec", keep: true},
+		"root/var/vcap/jobs-src/new_hostname/config_spec.json":    {desc: "new_hostname config spec", keep: true},
+		"root/var/vcap/packages/tor":                              {desc: "package symlink", typeflag: tar.TypeSymlink, keep: true},
+		"root/var/vcap/jobs-src/tor/job.MF":                       {desc: "job manifest file", typeflag: TypeMissing},
 	}
+	actual := make(map[string][]byte)
+
+	populator := roleImageBuilder.NewDockerPopulator(rolesManifest.Roles[0], releasePathConfigSpec)
+
+	pipeR, pipeW, err := os.Pipe()
+	assert.NoError(err, "Failed to create a pipe")
+
+	tarWriter := tar.NewWriter(pipeW)
+	tarReader := tar.NewReader(pipeR)
+	var asyncError error
+	mutex := sync.Mutex{}
+	go func() {
+		mutex.Lock()
+		defer mutex.Unlock()
+		defer tarWriter.Close()
+		asyncError = populator(tarWriter)
+	}()
 
-	symlinkPath := filepath.Join(dockerfileDir, "root/var/vcap/packages/tor")
-	if pathInfo, err := os.Lstat(symlinkPath); assert.NoError(err) {
-		assert.Equal(os.ModeSymlink, pathInfo.Mode()&os.ModeSymlink)
-		if target, err := os.Readlink(symlinkPath); assert.NoError(err) {
-			pkg := getPackage(rolesManifest.Roles, "myrole", "tor", "tor")
-			if assert.NotNil(pkg, "Failed to find package") {
-				expectedTarget := filepath.Join("..", "packages-src", pkg.Fingerprint)
-				assert.Equal(expectedTarget, target)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if !assert.NoError(err, "Error reading tar file") {
+			break
+		}
+		if info, ok := expected[header.Name]; ok {
+			delete(expected, header.Name)
+			if info.typeflag == tar.TypeRegA {
+				info.typeflag = tar.TypeReg
+			}
+			if header.Typeflag == tar.TypeRegA {
+				header.Typeflag = tar.TypeReg
+			}
+			if info.typeflag == TypeMissing {
+				assert.Fail("File %s should not exist", header.Name)
+				continue
+			}
+			assert.Equal(info.typeflag, header.Typeflag, "Unexpected type for item %s", header.Name)
+			if info.keep {
+				if header.Typeflag == tar.TypeSymlink || header.Typeflag == tar.TypeLink {
+					actual[header.Name] = []byte(header.Linkname)
+				} else {
+					buf := &bytes.Buffer{}
+					_, err = io.Copy(buf, tarReader)
+					assert.NoError(err, "Error reading contents of %s", header.Name)
+					actual[header.Name] = buf.Bytes()
+				}
 			}
 		}
 	}
-
-	// job.MF should not be there
-	assert.Error(util.ValidatePath(filepath.ToSlash(filepath.Join(dockerfileDir, "root/var/vcap/jobs-src/tor/job.MF")), false, "job manifest file"))
-
-	// And verify the config specs are as expected
-	jsonPath := filepath.Join(dockerfileDir, "root/var/vcap/jobs-src/new_hostname/config_spec.json")
-	buf, err := ioutil.ReadFile(jsonPath)
-	if !assert.NoError(err, "Failed to read new_hostname/config_spec.json %s\n", jsonPath) {
-		return
+	// Synchronize with the gofunc to make sure it's done
+	mutex.Lock()
```

Comment:
This is a race. You can't be sure anything has started in a separate thread. Use a channel latch

```go
latch := make(chan struct{})

go func() {
   close(latch) // I'm ready!
}()

<-latch // I won't be able to go until the close has occurred.
```